be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+{
+	/*TODO:Complete implementation during persistent list implementation */
+	struct m0_be_dtm0_log *log;
+
+	switch (op) {
+	case M0_DTML_CREATE:
+		M0_BE_ALLOC_CREDIT_PTR(log, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->dl_seg, seg, accum);
+		M0_BE_ALLOC_CREDIT_PTR(log->u.dl_list, seg, accum);
+		lrec_be_list_credit(M0_BLO_CREATE, 1, accum);
+		m0_be_tx_credit_add(accum, &M0_BE_TX_CREDIT_PTR(log));
+		break;
+	case M0_DTML_EXECUTED:
+	case M0_DTML_PERSISTENT:
+		dtm0_be_alloc_log_rec_credit(txd, payload, seg, accum);
+		dtm0_be_set_log_rec_credit(txd, payload, seg, accum);
+		m0_be_list_credit(M0_BLO_ADD, 1, accum);
+		break;
+	case M0_DTML_PRUNE:
+		dtm0_be_del_log_rec_credit(seg, rec, accum);
+		m0_be_list_credit(M0_BLO_DEL, 1, accum);
+		break;
+	case M0_DTML_REDO:
+	default:
+		M0_IMPOSSIBLE("");
+	}
+}
+
+M0_INTERNAL int m0_be_dtm0_log_create(struct m0_be_tx        *tx,
+                                      struct m0_be_seg       *seg,
+                                      struct m0_be_dtm0_log **out)
+{
+	struct m0_be_dtm0_log *log;
+
+	M0_PRE(tx != NULL);
+	M0_PRE(seg != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log, seg, tx);
+	M0_ASSERT(log != NULL);
+
+	M0_BE_ALLOC_PTR_SYNC(log->u.dl_list, seg, tx);
+	M0_ASSERT(log->u.dl_list != NULL);
+
+	log->dl_seg = seg;
+
+	lrec_be_list_create(log->u.dl_list, tx);
+	M0_BE_TX_CAPTURE_PTR(seg, tx, log);
+	*out = log;
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_destroy(struct m0_be_tx        *tx,
+                                        struct m0_be_dtm0_log **log)
+{
+#if 0 /* TODO: write down persistent implementation later */
+	struct m0_be_dtm0_log *plog = *log;
+
+	M0_PRE(plog != NULL);
+	m0_free(plog->u.dl_tlist);
+	m0_free(plog);
+	*log = NULL;
+#endif
+}
+
+M0_INTERNAL
+struct m0_dtm0_log_rec *m0_be_dtm0_log_find(struct m0_be_dtm0_log    *log,
+                                            const struct m0_dtm0_tid *id)
+{
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	if (log->dl_is_persistent) {
+		struct m0_dtm0_log_rec *lrec;
+
+		m0_be_list_for(lrec, log->u.dl_list, lrec) {
+			if (m0_dtm0_tid_cmp(log->dl_cs, &lrec->dlr_txd.dtd_id,
+                                          id) == M0_DTS_EQ) {
+				break;
+			}
+		} m0_be_list_endfor;
+		return lrec;
+	} else {
+		return m0_tl_find(lrec, rec, log->u.dl_tlist,
+			          m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id,
+                                          id) == M0_DTS_EQ);
+	}
+}
+
+static int be_dtm0_log_rec_init(struct m0_dtm0_log_rec **rec,
+				struct m0_be_tx         *tx,
+				struct m0_dtm0_tx_desc  *txd,
+				struct m0_buf           *payload)
+{
+	int                     rc;
+	struct m0_dtm0_log_rec *lrec;
+
+	M0_ALLOC_PTR(lrec);
+	if (lrec == NULL)
+		return M0_ERR(-ENOMEM);
+
+	rc = m0_dtm0_tx_desc_copy(txd, &lrec->dlr_txd);
+	if (rc != 0) {
+		m0_free(lrec);
+		return rc;
+	}
+
+	rc = m0_buf_copy(&lrec->dlr_payload, payload);
```

Comment:
The "m0_buf_copy(dest, src)" has dest as its first parameter, and src as its second parameter.
But, m0_dtm0_tx_desc_copy(src, dest) does the other way.

Please also refer to "strcpy()".
Maybe we need a more uniform convention.

Reply:
This will be reworked as a part of https://jts.seagate.com/browse/EOS-19332 in the next sprint, closing.